portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```
@@ -35,39 +37,55 @@
     /** GET request pulls comments from datastore and prints on /comments page */
     @Override
     public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-        Query query = new Query("Comment").addSort("timestamp", SortDirection.ASCENDING);
-
+        Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
         PreparedQuery results = datastore.prepare(query);
 
         // new arraylist everytime to keep thread safe
         ArrayList<String> comments = new ArrayList<>();
         
+        int requestedNumComments = getBoundedMaxComments(request);
+        System.out.println(requestedNumComments);
         for (Entity entity : results.asIterable()) {
-            String comment = (String) entity.getProperty("text");
-            comments.add(comment);
+            if (requestedNumComments > 0) {
+                String comment = (String) entity.getProperty("text");
+                comments.add(comment);
+                requestedNumComments--;
+            }
         }
 
+        // reverse order to oldest first
+        Collections.reverse(comments);
+
         // convert self object to json and print on /comment page
         Gson gson = new Gson();
         response.setContentType("application/json;");
         response.getWriter().println(gson.toJson(comments));
     }
 
+    /** Parses user-requested max number of comments, returned num is valid and positive */
+    private int getBoundedMaxComments(HttpServletRequest request) {
+        int requestedComments = 5;
+        int defaultComments = 5;
+        String inputMaxComments = request.getParameter("max-comments");
+        if (inputMaxComments != null) {
```

Interesting for the input to resolve to null if I'm assuming the input is left empty rather than an error? I had assumed it would just be an empty string. Can you confirm if that's the case, and catch it if null really is an error?

Also might want to catch the case where a non-numeric string is given. This can happen if say some user modifies the html on their browser to change the input's type from "number" to "text". (good to keep in mind that any and all exposed html can be corrupted by the user since they have full control of it)

---

yeah, I was getting a null error before. Will catch both those errors

---

ignoring the error if thrown, is that ok?